File: openrndr-adopted-artifex/src/main/java/io/lacuna/artifex/Ring2.java
package io.lacuna.artifex;

import io.lacuna.artifex.utils.Intersections;
import io.lacuna.bifurcan.LinearList;
import io.lacuna.bifurcan.Lists;

import java.util.Arrays;
import java.util.Comparator;

import static io.lacuna.artifex.Vec.vec;
import static io.lacuna.artifex.utils.Intersections.*;
import static io.lacuna.artifex.utils.Scalars.EPSILON;
import static java.lang.Math.abs;
import static java.lang.Math.signum;

public class Ring2 {

  public static class Result {
    public static final Result INSIDE = new Result(true, null);
    public static final Result OUTSIDE = new Result(false, null);

    public final boolean inside;
    public final Curve2 curve;

    private Result(boolean inside, Curve2 curve) {
      this.inside = inside;
      this.curve = curve;
    }

    public Result(Curve2 curve) {
      this(true, curve);
    }

    @Override
    public String toString() {
      if (inside) {
        return curve == null ? "INSIDE" : "EDGE: " + curve;
      } else {
        return "OUTSIDE";
      }
    }
  }

  public final Curve2[] curves;
  public final Box2 bounds;
  public final boolean isClockwise;
  public final double area;

  private Ring2(Curve2[] curves, Box2 bounds, boolean isClockwise, double area) {
    this.curves = curves;
    this.bounds = bounds;
    this.isClockwise = isClockwise;
    this.area = area;
  }

  public Ring2(Iterable<Curve2> cs) {

    // TODO: dedupe collinear adjacent lines
    Box2 bounds = Box2.EMPTY;
    double signedArea = 0;
    LinearList<Curve2> list = new LinearList<>();
    for (Curve2 a : cs) {
      for (Curve2 b : a.split(a.inflections())) {

        list.addLast(b);
        bounds = bounds.union(b.start()).union(b.end());
        signedArea += b.signedArea();
      }
    }

    this.isClockwise = signedArea < 0;
    this.area = abs(signedArea);
    this.bounds = bounds;

    curves = list.toArray(Curve2[]::new);
    for (int i = 0; i < curves.length - 1; i++) {
      curves[i] = curves[i].endpoints(curves[i].start(), curves[i + 1].start());
    }
    int lastIdx = curves.length - 1;
    curves[lastIdx] = curves[lastIdx].endpoints(curves[lastIdx].start(), curves[0].start());
  }

  public static Ring2 of(Curve2... cs) {
    return new Ring2(Lists.from(cs));
  }

  public Region2 region() {
    return new Region2(LinearList.of(this));
  }

  /**
   * @return a unit square from [0, 0] to [1, 1]
   */
  public static Ring2 square() {
    return Box.box(vec(0, 0), vec(1, 1)).outline();
  }

  /**
   * @return a unit circle with radius of 1, centered at [0, 0]
   */
  public static Ring2 circle() {
    // taken from http://whizkidtech.redprince.net/bezier/circle/kappa/
    double k = 4.0 / 3.0 * (StrictMath.sqrt(2) - 1);
    return Ring2.of(
      Bezier2.curve(vec(1, 0), vec(1, k), vec(k, 1), vec(0, 1)),
      Bezier2.curve(vec(0, 1), vec(-k, 1), vec(-1, k), vec(-1, 0)),
      Bezier2.curve(vec(-1, 0), vec(-1, -k), vec(-k, -1), vec(0, -1)),
      Bezier2.curve(vec(0, -1), vec(k, -1), vec(1, -k), vec(1, 0)));
  }

  public Path2 path() {
    return new Path2(this);
  }

  public Ring2 reverse() {
    return new Ring2(
      LinearList.from(Lists.reverse(Lists.lazyMap(Lists.from(curves), Curve2::reverse))).toArray(Curve2[]::new),
      bounds,
      !isClockwise,
      area);
  }

  public Result test(Vec2 p) {

    if (!bounds.expand(SPATIAL_EPSILON).contains(p)) {
      return Result.OUTSIDE;
    }

    Line2 ray = Line2.line(p, vec(bounds.ux + 1, p.y));
    int count = 0;

    // since our curves have been split at inflection points, there can only
    // be a single ray/curve intersection unless the curve is collinear
    for (Curve2 c : curves) {
      Box2 b = c.bounds();
      boolean flat = b.height() == 0;

      //System.out.println(p + " " + b + " " + c);

      // it's to our right
      if (p.x < b.lx) {
        // check if we intersect within [bottom, top)
        if (p.y >= b.ly && p.y < b.uy) {
          //System.out.println("right, incrementing");
          count++;
        }

        // we're inside the bounding box
      } else if (b.expand(vec(SPATIAL_EPSILON, 0)).contains(p)) {
        Vec2 i = Arrays.stream(lineCurve(ray, c))
          .map(v -> v.map(n -> Intersections.round(n, Intersections.PARAMETRIC_EPSILON)))
          .filter(PARAMETRIC_BOUNDS::contains)
          .min(Comparator.comparingDouble(v -> v.x))
          .orElse(null);

        if (i != null) {
          //System.out.println(i);
          if (i.x == 0) {
            return new Result(c);
          } else if (!flat && p.y < b.uy) {
            //System.out.println("intersected, incrementing");
            count++;
          }
        } else {
          //System.out.println("no intersection");
        }
      }
    }

    //System.out.println(count);

    return count % 2 == 1 ? Result.INSIDE : Result.OUTSIDE;
  }

  public Ring2 transform(Matrix3 m) {
    return new Ring2(() -> Arrays.stream(curves).map(c -> c.transform(m)).iterator());
  }
}

File: openrndr-kartifex/src/commonMain/kotlin/Ring2.kt
package org.openrndr.kartifex

import org.openrndr.kartifex.utils.Intersections
import kotlin.math.abs
import kotlin.math.sqrt


class Ring2 {
    class Result private constructor(val inside: Boolean, val curve: Curve2?) {

        constructor(curve: Curve2) : this(true, curve)

        override fun toString(): String {
            return if (inside) {
                if (curve == null) "INSIDE" else "EDGE: $curve"
            } else {
                "OUTSIDE"
            }
        }

        companion object {
            val INSIDE: Result = Result(true, null)
            val OUTSIDE: Result = Result(false, null)
        }
    }

    val curves: Array<Curve2>
    val bounds: Box2
    val isClockwise: Boolean
    val area: Double

    private constructor(
        curves: Array<Curve2>,
        bounds: Box2,
        isClockwise: Boolean,
        area: Double
    ) {
        this.curves = curves
        this.bounds = bounds
        this.isClockwise = isClockwise
        this.area = area
    }

    constructor(cs: Iterable<Curve2>) {
        // TODO: dedupe collinear adjacent lines
        var bounds: Box2 = Box2.EMPTY
        var signedArea = 0.0
        val list: ArrayDeque<Curve2> = ArrayDeque()
        for (a in cs) {
            for (b in a.split(a.inflections())) {
                list.addLast(b)
                bounds = bounds.union(b.start()).union(b.end())
                signedArea += b.signedArea()
            }
        }
        isClockwise = signedArea < 0
        area = abs(signedArea)
        this.bounds = bounds
        curves = list.toTypedArray()
        for (i in 0 until curves.size - 1) {
            curves[i] = curves[i].endpoints(curves[i].start(), curves[i + 1].start())
        }
        val lastIdx = curves.size - 1
        curves[lastIdx] = curves[lastIdx].endpoints(curves[lastIdx].start(), curves[0].start())
    }

    fun region(): Region2 {
        return Region2(listOf(this))
    }

    fun test(p: Vec2): Result {
        if (!bounds.expand(Intersections.SPATIAL_EPSILON).contains(p)) {
            return Result.OUTSIDE
        }
        val ray: Line2 = Line2.line(p, Vec2(bounds.ux + 1, p.y))
        var count = 0

        // since our curves have been split at inflection points, there can only
        // be a single ray/curve intersection unless the curve is collinear
        for (c in curves) {
            val b: Box2 = c.bounds()
            val flat = b.height() == 0.0

            //System.out.println(p + " " + b + " " + c);

            // it's to our right
            if (p.x < b.lx) {
                // check if we intersect within [bottom, top)
                if (p.y >= b.ly && p.y < b.uy) {
                    //System.out.println("right, incrementing");
                    count++
                }

                // we're inside the bounding box
            } else if (b.expand(Vec2(Intersections.SPATIAL_EPSILON, 0.0))
                    .contains(p)
            ) {
                val i: Vec2? =
                    Intersections.lineCurve(ray, c)
                        .map { v: Vec2 ->
                            v.map { n: Double ->
                                Intersections.round(
                                    n,
                                    Intersections.PARAMETRIC_EPSILON
                                )
                            }
                        }
                        .filter { v ->  Intersections.PARAMETRIC_BOUNDS.contains(v) }
                        .minByOrNull { v: Vec2 -> v.x }

                if (i != null) {
                    //System.out.println(i);
                    if (i.x == 0.0) {
                        return Result(c)
                    } else if (!flat && p.y < b.uy) {
                        //System.out.println("intersected, incrementing");
                        count++
                    }
                } else {
                    //System.out.println("no intersection");
                }
            }
        }

        //System.out.println(count);
        return if (count % 2 == 1) Result.INSIDE else Result.OUTSIDE
    }

    fun transform(m: Matrix3): Ring2 = Ring2(curves.map { c -> c.transform(m) })

    companion object {
        fun of(vararg cs: Curve2): Ring2 {
            return Ring2(cs.toList())
        }

        /**
         * @return a unit square from [0, 0] to [1, 1]
         */
        fun square(): Ring2 {
            return Box.box(Vec2(0.0, 0.0), Vec2(1.0, 1.0))
                .outline()
        }

        /**
         * @return a unit circle with radius of 1, centered at [0, 0]
         */
        fun circle(): Ring2 {
            // taken from http://whizkidtech.redprince.net/bezier/circle/kappa/
            val k: Double = 4.0 / 3.0 * (sqrt(2.0) - 1)
            return of(
                Bezier2.curve(
                    Vec2(1.0, 0.0),
                    Vec2(1.0, k),
                    Vec2(k, 1.0),
                    Vec2(0.0, 1.0)
                ),
                Bezier2.curve(
                    Vec2(0.0, 1.0),
                    Vec2(-k, 1.0),
                    Vec2(-1.0, k),
                    Vec2(-1.0, 0.0)
                ),
                Bezier2.curve(
                    Vec2(-1.0, 0.0),
                    Vec2(-1.0, -k),
                    Vec2(-k, -1.0),
                    Vec2(0.0, -1.0)
                ),
                Bezier2.curve(
                    Vec2(0.0, -1.0),
                    Vec2(k, -1.0),
                    Vec2(1.0, -k),
                    Vec2(1.0, 0.0)
                )
            )
        }
    }
}